input/kafka/inputkafka.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	return &conf, nil
+}
+
+// Start wraps the actual function starting the plugin
+func (t *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEvent) (err error) {
+	/**
+	 * Setup a new Sarama consumer group
+	 */
+	cum := consumerHandle{
+		i:     t,
+		ch:    msgChan,
+		ready: make(chan bool),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	client, err := sarama.NewConsumerGroup(t.Brokers, t.Group, t.saConf)
+	if err != nil {
+		goglog.Logger.Errorf("Error creating consumer group client: %v", err)
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, t.Topics, &cum); err != nil {
+				goglog.Logger.Errorf("Error from consumer: %v", err)
+			}
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				return
+			}
+
+			cum.ready = make(chan bool)
+		}
+	}()
+
+	<-cum.ready // Await till the consumer has been set up
+	goglog.Logger.Println("Sarama consumer up and running!...")
+
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, syscall.SIGINT, syscall.SIGTERM)
+	select {
+	case <-ctx.Done():
+		goglog.Logger.Println("terminating: context cancelled")
+	case <-sigterm:
+		goglog.Logger.Println("terminating: via signal")
+	}
+	cancel()
+	wg.Wait()
+	if err = client.Close(); err != nil {
+		goglog.Logger.Errorf("Error closing client: %v", err)
+		return err
+	}
+	return nil
+}
+
+// consumerHandle represents a Sarama consumer group consumer
+type consumerHandle struct {
+	i     *InputConfig
+	ch    chan<- logevent.LogEvent
+	ready chan bool
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (c *consumerHandle) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	c.ready <- true
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (c *consumerHandle) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (c *consumerHandle) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+
+	// NOTE:
+	// Do not move the code below to a goroutine.
+	// The `ConsumeClaim` itself is called within a goroutine, see:
+	// https://github.com/Shopify/sarama/blob/master/consumer_group.go#L27-L29
+	for message := range claim.Messages() {
+		//goglog.Logger.Printf("Message claimed: value = %s, timestamp = %v, topic = %s", string(message.Value), message.Timestamp, message.Topic)
+		var extra = map[string]interface{}{
+			"topic": message.Topic,
+			"timestamp": message.Timestamp,
+		}
+		ok, err := c.i.Codec.Decode(context.TODO(), string(message.Value), extra, []string{}, c.ch)
```

Comment:
Why use `context.TODO()`?
1. why not main context?
2. why not `context.Background()`?

Reply:
changed to main context